### DISCUSSION TITLE
Database binding to app—good pratice

### DISCUSSION THREAD

**Discussion by MistaIA** [fastapi/fastapi#10558] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uvicorn
from fastapi import FastAPI
from motor.motor_asyncio import AsyncIOMotorClient
from pydantic import Field, EmailStr
from beanie import init_beanie, Document

# configs is a pydantic_settings loading .env
from Config import configs


class Sample(Document):
    name: str = Field(...)
    email: EmailStr = Field(...)
    course: str = Field(...)
    gpa: float = Field(..., le=4.0)

    class Settings:
        name = 'sample'


class AsyncDatabase:
    def __init__(self):
        self.async_engine = None
        self.client: AsyncIOMotorClient = AsyncIOMotorClient(
            f'{configs.DB_SCHEMA}://{configs.DB_USERNAME}:{configs.DB_PASSWORD}@{configs.DB_HOST}/?retryWrites=true&w=majority')

    async def set_engine(self):
        self.async_engine = await init_beanie(
            database=self.client[configs.DB_NAME],
            document_models=[Sample])


async_db: AsyncDatabase = AsyncDatabase()


class EngineManager:
    async def initiate(self, app: FastAPI):
        await async_db.set_engine()
        app.state.db = async_db
        await app.state.db.async_engine.begin()

    async def close(self, app: FastAPI):
        async_db.client.close()
        await app.state.db.async_engine.dispose()


db = EngineManager()


# Events
def execute_backend_server_event_handler(backend_app: FastAPI):
    async def launch_backend_server_events() -> None:
        await db.initiate(app=backend_app)

    return launch_backend_server_events


def terminate_backend_server_event_handler(backend_app: FastAPI):
    async def stop_backend_server_events() -> None:
        await db.close(app=backend_app)

    return stop_backend_server_events


def initialize_backend_application() -> FastAPI:
    app = FastAPI()

    app.add_event_handler(
        'startup',
        execute_backend_server_event_handler(backend_app=app),
    )
    app.add_event_handler(
        'shutdown',
        terminate_backend_server_event_handler(backend_app=app),
    )

    return app


fast_app: FastAPI = initialize_backend_application()


if __name__ == '__main__':
    uvicorn.run(
        app='test:fast_app',
        host=configs.SERVER_HOST,
        port=configs.SERVER_PORT,
        reload=configs.DEBUG,
        workers=configs.SERVER_WORKERS,
        log_level=configs.LOGGING_LEVEL,
    )
```

### Description

Using FastAPI + beanie (or any better MongoDB engine), This outputs : `AttributeError: 'NoneType' object has no attribute 'begin'`.
Which if I understood, means that `init_beanie` does not return something, so that FastAPI could do `begin()`, right?

I was not able to find `state` in the FastAPI documentation but by reading the source code, it comes from `Starlette` inheritance even though PyCharm highlights it as `Unresolve attribute reference`. I would appreciate materials about that. Therefore, is it a good practice, why and how to bind db to fast app?

Related questions (Some explanation there might be helpful): https://github.com/tiangolo/fastapi/discussions/10083#discussion-5522634
Material used: https://github.com/Aeternalis-Ingenium/FastAPI-Backend-Template/blob/trunk/backend/src/main.py

### Operating System

Windows

### Operating System Details

11

### FastAPI Version

0.104.0

### Pydantic Version

2.4.2

### Python Version

3.11.6

### Additional Context

```text
beanie==1.23.1
motor==3.3.1
starlette==0.27.0
starlette-exporter==0.16.0
uvicorn==0.23.2
```

**Comment by grillazz** [comment]
some of theres issues solved here https://github.com/grillazz/fastapi-mongodb

**Comment by YuriiMotov** [comment]
You can use lifespan and store db instance in state.

Docs:
* https://fastapi.tiangolo.com/advanced/events/
* https://www.starlette.io/applications/#storing-state-on-the-app-instance
* https://www.starlette.io/lifespan/
